### ISSUE TITLE
Fish autocomplete sometimes forgets history and goes back in time

### ISSUE THREAD

**Issue by jacobsa** [fish-shell/fish-shell#10300]
I'm using fish 3.6.4 with [this configuration](https://github.com/jacobsa/dotfiles/blob/linux/config.fish). Typically I run fish within tmux, with a few different tmux windows open at a time.

I find that a few times per day, fish seems to forget history for commands I run all of the time. For example, I run a command of this form many times per day:

```
foo; and bar; and baz
```

Normally when I typ `fo`, fish will automatically complete the rest of it and I can press right, then enter, to run it. But a few times per day _one instance_ of fish will totally forget it, and auto-complete a broken typo version I ran months ago:

```
foo; and basdfg; and baz
```

If I switch to other instances of fish running in other tmux windows, they don't have this problem: they still autocomplete correctly. If I close the tmux window and open a new one, the new instance of fish is fine. Even if I run `fish` to get a new sub-shell within the broken instance, it also uses the correct history. It's only the one busted instance of fish.

To be clear, it's not just one command. This issue happens with several different commands that I run frequently, and it seems like when fish gets into this state it's _all_ recent history that is lost, not just one command.

It's almost as if there is some in-memory circular buffer of history, and it eventually overflows when I run enough commands, and then fish is confused and thinks old history is current.

How can I debug this issue? Is there anything I can collect to help you understand it better?

---

```
> fish --version
fish, version 3.6.4

> uname -a
Linux <redacted>.com 6.5.13-1rodete2-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.5.13-1rodete2 (2024-01-11) x86_64 GNU/Linux

> echo $TERM
screen-256color

> tmux -V
tmux 3.3a
```

**Comment by krobelus** [comment]
Is this on NFS?
When it happens, is there a file like `~/.local/share/fish/fish_history.2NsTOE` that matches the history of the broken session? To compare, maybe dump the entire history with `history`.
Does `history merge` unbreak it?

**Comment by jacobsa** [comment]
It's not NFS, no. There is currently no file in `~/.local/share/fish/` but `fish_history` itself, but I'm not sure if it's expected to persist after the broken session.

What does `history merge` do? I can try it next time I see the problem.

**Comment by krobelus** [comment]
> What does history merge do? I can try it next time I see the problem.

It incorporates new history from other, newer shells; the crucial side effect is that it will reload the history file.
If it doesn't, you might need to run some command in another shell first, so `history merge` will actually do something.

**Comment by jacobsa** [comment]
Okay two updates:

*  If I run `history | grep foo | head` when a shell is in this state, it also agrees that the history is busted. It shows only the old typo version, or at least the more recent version doesn't show up in the first ten entries.

*  `history merge` does seem to restore the shell to a working state.

Any ideas what is causing the shell to get into this state in the first place? There is no network file system involved, and it's not like it _starts_ with bad history; it just becomes bad eventually.

**Comment by faho** [comment]
What does your system clock say, i.e. `date`?

**Comment by jacobsa** [comment]
Definitely from previous shells; these are commands I run every day and have done so for years. I _think_ they usually were probably also run in the current shell before they were forgotten, but I'm less sure about that.

`date` shows the correct current local time, in both the busted shell and a new one.

**Comment by krobelus** [comment]
This is a really mysterious bug, thanks for taking the effort to report. Maybe there's a simple explanation like you set `fish_history` or `fish_private_mode`.. but those can't be it because `history merge` works.

When we create a history, or run `history merge`, we record the current system time.
Every command from the history file that has a later timestamp will be excluded, because those commands are assumed to come from other shell instances.

Maybe there is some memory corruption that stomps the timestamp, setting it to some time in the past. That would explain why fish suddendly starts ignoring those.

We probably need to add some logging.

Can't hurt to try latest Git master as well, although I don't expect a change.

**Comment by jacobsa** [comment]
Thanks for being attentive to my weird bug. :-) If you want to add some logging and tell me how to check it, I'd be more than happy to report back.

---

There was one other theory I had, as someone who doesn't know fish's internals. How well is the plain text history format able to "re-synchronize" after corruption? Like if a line gets deleted or messed up or duplicated or something, are all of the entries after that liable to not work correctly? Or does it re-synchronize by looking for some marker afterward? I have in the past manually deleted entries from the file, and I worry if the format isn't robust it may have left me with a persistent problem if I made any editing mistakes.

**Comment by zanchey** [comment]
Running with `fish --debug history,history-file` might help.

**Comment by jacobsa** [comment]
Thanks, can you help me understand what that does? Will there be debug output somewhere that I can examine, and what am I looking for?

**Comment by faho** [comment]
> Will there be debug output somewhere that I can examine, and what am I looking for?

Yes, there will be debug output. By default it'll go to the terminal, add `--debug-output /path/to/file` to write it to a file.

I don't think we currently know anything specific to be looking for.

**Comment by magnetophon** [comment]
I just wanted to mention I have the exact same symptoms.
Let me know if there's anything I can do to help.

**Comment by magnetophon** [comment]
I also sometimes delete stuff from my history, just thought i'd mention it.

> What is the output of history when this happens? Do the offending entries turn up there?

I'll report back the next time it happens.

> Is your system clock okay? Does it move?

Never noticed any issues.

> Do you have any configuration? 

Yes:
https://github.com/magnetophon/.dot/tree/master/common/.config/fish

> Especially anything history-related, but also e.g. the fzf integration?

I use a couple of plugins: 
```
jorgebucaran/fisher
ilancosman/tide@v6
franciscolourenco/done
patrickf1/fzf.fish
meaningful-ooo/sponge
decors/fish-colored-man
jorgebucaran/autopair.fish
lewisacidic/fish-git-abbr
```
sponge and fzf.fish are the history related ones.

> Does this happen with 3.7.0?

Yes it does.

> Do you have any network mounts?

No.

> Do you suspend your system?

Yes I do.

> Does fish --debug history'*' --debug-output /tmp/history.txt show anything sensible if this happened? Unfortunately you'll have to run fish with that explicitly.

I have fish as my system shell in NixOS.
I'd have to figure out how to make the above command my default shell.

**Comment by faho** [comment]
> it forgot several commands, all of which are commands I ran in other terminals while that one was open

Fish does not read the history from other sessions after startup unless you run `history merge`, so that is as expected.

**Comment by 0ion9** [comment]
I think I mentioned this bug to krobelus at some point (few years ago?) but did not report. FWIW it is also still happening for me, and I can confirm that the commands in question are 'old commands that were in the history' rather than items recently added by work in other presently-running instances of fish.

List of details relating to questions asked thus far:
* fish-git 3.7.1.g3af849d73
* terminal: Kitty 0.33.0-1
* OS: Arch Linux x86_64
* not using tmux or any other multiplexer
* I do not use suspend or NFS mounts (but do rarely use FUSE mounts, eg `archivemount`); in any case the history file does not have any relation to these in my case
*  I do use ntpd, but as far as I can tell my system clock keeps reliable time, so it shouldn't be incurring large adjustments. 
* `history` is afflicted with forgetfulness in an identical way to autocomplete.
* I do (rarely) delete history items, but only using the `history delete` interface
* I don't set `fish_history` or use private mode
* IIRC I have the fzf integration enabled, but don't use it to interact with history
* Filesystem of /home/: ext4, in case that makes any difference to obscure timestamp interaction type things.

* "It's almost as if there is some in-memory circular buffer of history, and it eventually overflows when I run enough commands, and then fish is confused and thinks old history is current." -- I think this is incorrect, at least regarding the problem as it manifests on my system. Often, the problem will occur in a nominally 'fresh' shell (kitty will automatically start several shell instances, and when I eventually switch to the next instance, nothing has been done in that instance at that point but it is clearly afflicted with forgetfulness)

* I have not so far checked the output of `date` in an afflicted shell instance, but I can do that
* Will try to get logging setup.

**Comment by krobelus** [comment]
someone needs to look at the code and add relevant logs

**Comment by planet36** [comment]
I've been increasingly encountering a problem like this one for the past few months.
If manifests after I first log in.  When I type `history`, no history is printed, even though `fish_history` still exists and contains earlier history.

The last time it happened, I noticed there were embedded NUL bytes in fish_history.  (screenshot included)
When I removed the entry containing the NUL bytes, and logged out & in, things worked normally.


$ fish --version
> fish, version 3.7.1

$ uname -srm
> Linux 6.8.9-arch1-1 x86_64

$ fisher list
> jorgebucaran/fisher
> jethrokuan/z

$ tmux -V
> tmux 3.4

`/` is ext4

$ wc --lines fish_history.BAD
>141748 fish_history.BAD

![Clipboard_05-05-2024_01](https://github.com/fish-shell/fish-shell/assets/387896/6df04f79-9c27-4d22-9a0d-06585d36ac13)

**Comment by jacobsa** [comment]
@planet36 Maybe best to file a different bug for that, so as to avoid derailing this one? It seems like different symptoms (of course the cause could be related in the end).

**Comment by planet36** [comment]
> @planet36 Maybe best to file a different bug for that, so as to avoid derailing this one? It seems like different symptoms (of course the cause could be related in the end).

I just created this issue [10483](https://github.com/fish-shell/fish-shell/issues/10483)

**Comment by jeremyFreeAgent** [comment]
I have the same issue with version 3.7.1

`history` return the correct history but when typing, it return not the last entry but the previous one.

**Comment by jacobsa** [comment]
This is happening to me very frequently lately, with a particular command being forgotten. Is there any sort of useful debugging information I could collect about this while it's reproducing frequently? How can we move this forward?

**Comment by ybc37** [comment]
@jacobsa 

> [...] Is there any sort of useful debugging information I could collect about this while it's reproducing frequently? [...]

See https://github.com/fish-shell/fish-shell/issues/10300#issuecomment-1983921556 for getting debug information.

**Comment by jacobsa** [comment]
Thanks, but several people have already provided that output. So correct me if I'm wrong, but it seems that wasn't enough to help narrow this down. How can we move this forward?

**Comment by devsnek** [comment]
I'm getting the same issue with null bytes in the history file. It has happened to me three times in the last couple months and I'm not sure what I did to cause it.

**Comment by planet36** [comment]
> I'm getting the same issue with null bytes in the history file. It has happened to me three times in the last couple months and I'm not sure what I did to cause it.

Do you have anything in common with my usage/environment? https://github.com/fish-shell/fish-shell/issues/10300#issuecomment-2094911345

**Comment by devsnek** [comment]
I also use arch btw but I'm not using fisher or tmux.

One thing I noticed in the commit history is this change in january which touches code that happens to include logic for filling trailing bytes in the mapped history with null... it could very well be a red herring though. https://github.com/fish-shell/fish-shell/commit/f3e8272c5dab769b0e1afdea120a6641f9e74685#diff-c94c930db80889c1596fda236ca6258864d72425e9820a20820cb7c9bef54d33

**Comment by planet36** [comment]
I've been running fish with these debug options:
`--debug history'*' --debug-output ~/fish-debug-history.$(date +'%s').log`

I've noticed that when the problem of embedded NUL bytes happens, I see something like the following in the history log file.

```
history: Loaded 21348 old items
history: Saving 1 items via appending
history: Saving 1 items via appending
history: Loaded 21346 old items
history: Saving 1 items via appending
history: Saving 1 items via rewrite
history: Loaded 21340 old items
history: Saving 1 items via appending
history: Saving 1 items via appending
history: Saving 1 items via appending
```

or

```
history: Loaded 21347 old items
history: Saving 1 items via appending
history: Saving 1 items via appending
history: Loaded 21345 old items
history: Saving 1 items via rewrite
history: Loaded 21343 old items
history: Saving 1 items via appending
history: Saving 1 items via appending
history: Saving 1 items via appending
history: Saving 1 items via appending
history: Saving 1 items via appending
history: Saving 1 items via appending
```

Whereas when there's no bug, the history log file has a single `history: Loaded` line.

**Comment by shamilovtim** [comment]
Also experiencing this. My history is completely bricked and not saving. This is despite the history file existing and being full of history.

**Comment by kjnsn** [comment]
Looks like this would be fixed by https://github.com/fish-shell/fish-shell/pull/10680, as invalid lines would be simply ignored. IMHO shell history does not need to be absolutely correct, it's okay to drop one or two entries if something happened (like power loss) and keep what is possible.

**Comment by Xiretza** [comment]
That's a workaround at best, the history file should not be getting corrupted in the first place.

**Comment by kjnsn** [comment]
> That's a workaround at best, the history file should not be getting corrupted in the first place.

Oh absolutely. Unfortunately there are some situations where it's unavoidable, such as power loss on a non-journaled filesystem. 

Also see https://www.sciencedirect.com/topics/computer-science/graceful-degradation.

**Comment by devsnek** [comment]
this issue is not about power loss though. none of the occurrences I've had of this issue involved power loss.

**Comment by jacobsa** [comment]
It's hard to know for sure none of your issues have involved power loss, unless you're positive that's never happened in the entire time the file existed. A workaround would be nice, allowing us not to have to suffer from this problem and making the question of _why_ corruption has happened an orthogonal issue that can be investigated separately.

**Comment by mortbauer** [comment]
I'm also having this issue every now and then, just had it right now, and it was in that case a corrupted history file as pointed out earlier. I think it wasn't any power issue or hang I think (though I had a complete system freeze yesterday, but I'm quite sure history worked still after that one).
Anyways manually fixing the file and running history merge and everything is fine again.
Here is the screenshot of my corruption:
![image](https://github.com/user-attachments/assets/6f91102e-b347-49d8-a42d-5ef0ab298f2f)

**Comment by jacobsa** [comment]
I see, thanks. I was just going off of the symptoms I saw (from the original post), where it's impossible to tell when things "went wrong" because the problem just happens eventually. If you've narrowed it down to something more specific like newly-added null bytes, that's great.

**Comment by kjnsn** [comment]
Sorry I must have really misled people with "power loss". I should have said "any event that corrupts the history file"

**Comment by planet36** [comment]
If anyone has history debugging turned on, do you notice more than one "Loaded #### old items" in the log when the history file is corrupted?

Examples:
https://github.com/fish-shell/fish-shell/issues/10300#issuecomment-1986726514
https://github.com/fish-shell/fish-shell/issues/10300#issuecomment-2230841434

I noticed that subsequent "Loaded" lines have a lesser value than previous ones.  That seemed counter-intuitive to me.

**Comment by krobelus** [comment]
I pushed the workaround suggested by @kjnsn to drop corrupted (= NUL-prefixed) history entries. Earlier history entries will now be loaded successfully again. We already dropped corrupted history entries before, but only whenever we vacuum the history file (which happens on average once every 25 commands).

If you can reproduce this, please run with `fish -d history_file` to print any unexpected errors on stderr.

If ou hit this again, please run the command below and LMK if this fails to find a corrupted line (because then we have a separate issue).

```
 perl -ne 'print "CORRUPTED: $_" if m/\0/' < $__fish_data_dir/(set -q fish_history && echo $fish_history || echo fish)_history
```

The leading space is important: otherwise fish might vacuum the history file, which removes any corrupted entries.


If you're not running latest master, fish will vacuum despite the leading space, so in that case prefer to run an equivalent command in another shell (like bash), without running an interactive command in fish that might vacuum the history file.

Also if you hit this on latest master, you'll get a default-enabled `ignoring corrupted history entry around offset` message on stderr.

**Comment by krobelus** [comment]
I'd call it done until someone reproduces it or wants to do more work on it

**Comment by krobelus** [comment]
Sorry there was a bug in that version that would cause such breakage. Need to use latest master

**Comment by radupotop** [comment]
Apologies for the bump. I'm seeing the same issue, with my shell history completely disappearing. I'm usually seeing this issue when I open multiple terminals at the same time.  
I currently have two terminals open, with both of them showing very short and different histories from the main one.

But also, I see this:
```
❯ ll -a .local/share/fish/
total 4.7M
drwx------  3 tail users 4.0K Oct 21 11:30 ./
drwx------ 92 tail users 4.0K Oct 18 15:34 ../
-rw-r--r--  1 tail users 4.5M Oct 21 11:30 fish_history
-rw-------  1 tail users    0 Oct 21 11:30 fish_history.cpvhpt
```

As soon as that `fish_history.cpvhpt` disappeared the history started working again as expected.

If I can help in any way with debugging this please let me know.

**Comment by jbrodriguez** [comment]
man, it's so reassuring to see i'm not alone, i started using fish 10 years ago (https://jbrio.net/posts/switching-from-zsh-to-fish/), never looked back, then this issue started hitting, it got a bit frustrating so i started using atuin, which is nice, but it's not as smooth as fish suggestions

in my case, it manifested like this 
- i start a terminal
- switch to a folder
- type `w`, it suggests ails init, `wails init` is a command i haven't used in months !
- what i've used more recently in this folder and i what i really intend to use is `watchman ....`

something similar happened for different commands

**Comment by sl4v3** [comment]
I'm also having fish shell forget commands.  This will happen in just one of several open terminals.  I'm using 3.71.  I'm pretty much using stock fish with tide installed.  It happens about once every other day.  history --merge will restore the missing command.

**Comment by krobelus** [comment]
@retrotails thanks. Did the affected session print any error about `history-file`? (those logs are enabled by default now)
I assume it did not, since you didn't mention it. That's quite surprising.
If I could get a debugger attached to an affected session (using a `CMAKE_BUILD_TYPE=Debug` build) that might help.
Alternatively, add more logs since apparently the `history_file` ones don't cover everything.
I checked that all syscalls either restart or print an error on EINTR but I might have missed one.

**Comment by devsnek** [comment]
this just happened to me again. `fish, version 3.7.1`. i also want to note its very consistent, its always a totally valid history entry prefixed with null characters.

**Comment by joanbm** [comment]
Some of the cases where NUL bytes appear in the history hint at it likely being related to unclean filesystem unmounts, as they show shutdown/reboot commands around the corrupted entries: https://github.com/fish-shell/fish-shell/issues/10300#issuecomment-2094911345, #10740

After some investigation, I found that *in ext4, in the default settings, if the filesystem is not cleanly unmounted, a `write` to a file opened with `O_APPEND` can appear as the file having the 'new' size but the data showing up as NUL bytes*. This is documented in [Specifying and Checking File System Crash-Consistency Models](https://www.comp.nus.edu.sg/~lijl/papers/ferrite-asplos16.pdf) (keyword: "prefix-append") and [ChromeOS Bug#172227346](https://issuetracker.google.com/issues/172227346). I have been able to reproduce this on current kernels with [this code](https://gist.github.com/joanbm/159759c075b04a679593ee5473456da0). I think that this potentially explains some of the cases in this thread.

(Bonus: In btrfs, if you rename over an existing file without calling fsync on the source file first - which as far as I can tell, is how fish does it when vacuuming the history file - [there's no guarantee that the target file is complete](https://archive.kernel.org/oldwiki/btrfs.wiki.kernel.org/index.php/FAQ.html#What_are_the_crash_guarantees_of_overwrite-by-rename.3F). I did some synthetic tests and it is at least possible that the target file is empty.

**Comment by magnetophon** [comment]
@joanbm FWIF, I'm using zfs, does that have similar issues?

**Comment by jbrodriguez** [comment]
this has happened to me on apfs, so not sure it's file system related

**Comment by joanbm** [comment]
@magnetophon I wasn't able to find any source online, but I did some quick tests and I wasn't able to get any trailing NUL bytes after an unclean shutdown on ZFS.

@jbrodriguez However, according to the paper above, other filesystems (e.g. f2fs) have similar behaviour - it's not something exclusive to ext4.

Also note that I am not sure how many of the cases in this issue my possibility explains - likely not all of them. There are only a few comments that have evidence of the history containing embedded NUL bytes (and all of them are potentially consistent with the possibility I mention). But it's not clear how many of the other reports in this thread are caused by embedded NUL bytes, or whether they are caused by something else.

**Comment by hamoid** [comment]
In my case I'm on a laptop, never power loss. The disk is an encrypted SSD with `ext4    defaults,noatime,data=ordered 0 1` in fstab.

In my case things go back to normal by deleting those null characters. As soon as I save the file, history works (no reboot needed).

**Comment by planet36** [comment]
I just encountered the problem of null bytes in the fish history file using version `4.0.0`.

I've been using this command to start fish with debug history enabled.
`fish --debug history'*' --debug-output ~/fish-debug-history.$(date +'%s').log`

The file that was generated when the crash happened has these contents.
> error: ignoring corrupted history entry around offset 2998198
> history: Loaded 23465 old items
> history: Saving 1 items via appending
> history: Saving 1 items via appending
> history: Saving 1 items via appending
> history: Saving 1 items via appending
> history: Saving 1 items via appending

Using gdb, the backtrace shows the following:

```
#0  __GI_abort () at abort.c:91
#1  0x00006480945a51aa in std::sys::pal::unix::abort_internal ()
#2  0x000064809435cca9 in std::process::abort ()
#3  0x0000648094595b70 in fish::panic::panic_handler::{{closure}} ()
#4  0x00006480945a4ef4 in std::panicking::rust_panic_with_hook ()
#5  0x00006480945d3255 in std::panicking::begin_panic_handler::{{closure}} ()
#6  0x00006480945d31e9 in std::sys::backtrace::__rust_end_short_backtrace ()
#7  0x00006480945d37fc in rust_begin_unwind ()
#8  0x00006480943502ff in core::panicking::panic_fmt ()
#9  0x000064809435034b in core::panicking::panic ()
#10 0x000064809457ea3d in fish::history::file::HistoryFileContents::decode_item ()
#11 0x00006480945095b1 in fish::history::HistoryImpl::item_at_index ()
#12 0x000064809451056d in fish::history::History::item_at_index ()
#13 0x000064809450fdfa in fish::history::HistorySearch::go_to_next_match ()
#14 0x0000648094514d16 in core::ops::function::FnOnce::call_once{{vtable.shim}} ()
#15 0x0000648094514845 in core::ops::function::FnOnce::call_once{{vtable.shim}} ()
#16 0x0000648094541801 in std::sys::backtrace::__rust_begin_short_backtrace ()
#17 0x000064809456cb2e in core::ops::function::FnOnce::call_once{{vtable.shim}} ()
#18 0x00006480945d3b3b in std::sys::pal::unix::thread::Thread::new::thread_start ()
#19 0x000072fc6ceba70a in start_thread (arg=<optimized out>) at pthread_create.c:448
#20 0x000072fc6cf3eaac in __GI___clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78
```

**Comment by planet36** [comment]
The `fish` binary I'm using is the one packaged by Arch Linux.

**Comment by planet36** [comment]
Contents of the debug history file from a crash that just happened:
> error: ignoring corrupted history entry around offset 2999525
> history: Loaded 23481 old items
> history: Saving 1 items via appending
> history: Saving 1 items via appending
> history: Saving 1 items via appending

The backtrace was the same (except for addresses) as in my earlier comment.  https://github.com/fish-shell/fish-shell/issues/10300#issuecomment-2712113050

**Comment by danielrainer** [comment]
I just created #11330, which contains small changes to how the history file is written. With this version, I have not experienced any history corruption for a while, which used to occur quite frequently for me before using this patch.

I don't know if these changes actually fix the corruption issue entirely, but if you have the same problem, it might be worth trying these changes for yourself and reporting your results. (Note that the changes are on the `history_flush` branch, not `master`, in my repo.

**Comment by marcecj** [comment]
> EDIT: In my case (fish 4.0-b1, btrfs), I am getting history files with NULs at the end increasingly often - without power loss. I even managed to reproduce it a 'live' crash as I was getting attempting to debug it!

I just wanted to add that I've been seeing this for the past year, maybe longer, and much like the OP, it doesn't happen often, but when it does, it's sudden, and AFAICT always when I launch a new shell.  The difference to OP is I generally use Konsole tabs, so what happens for me is: I open a new Konsole tab, start typing a (oft-run) command, expecting it to auto-complete, only it doesn't.  Whenever this happens, the NULs are never *right* at the end, but always a few entries before the end of the history file.  I generally manually delete them, after which history works as expected again.

(I use fish 3.7.1 on OpenSuse TW, using BTRFS; it just happened again on my laptop, so I decided to see if this was already reported, thus finding this issue.)

**Comment by marcecj** [comment]
Indeed, I do not use `history merge` (automatically *or* manually).

**Comment by devsnek** [comment]
I've been noticing this corruption a lot more often recently (ty to whoever made it log when corruption is detected), and it seems to always happen when I have multiple terminals open.

**Comment by danielrainer** [comment]
In the history related code, I noticed that the history is kept in a `mmap`ed file-backed buffer. The mapping happens here:
https://github.com/fish-shell/fish-shell/blob/68a8ab9667e08adc36d52baa9e0617441ad91e00/src/history/file.rs#L48-L57

According to the Linux 6.13 `mmap(2)` man page, when using `MAP_PRIVATE`:

> It  is  unspecified whether changes made to the file after the `mmap()` call are visible in the mapped region.

Since changes to the history file can definitely happen, this might be worth investigating. I have not checked what Linux (or any other kernel) actually does when a privately mapped file is modified, but it is at least something we can test and should account for regardless of the current behavior of certain kernels.

**Comment by joanbm** [comment]
I still get areas of 200+ NUL bytes once in a while on Arch+btrfs+SSD, without power loss. I have looking for a way to reproduce the issue (e.g. by having some tests that brute forces renames/appends/mmaps similar to fish) but found nothing interesting so far.

I found it very interesting that #11186 has a consistent reproducer for this problem *on zfs*, and it still reproduces on the latest master commit (which has #11330), on Arch+ZFS+NVMe. I have been taking a look, and it looks like only a single history entry gets replaced with NULs, and adding some logs it seems like the entry gets successfully [written via append](https://github.com/fish-shell/fish-shell/blob/692b53ab853e1221a5edadbe3fdf7afc37bd1329/src/history.rs#L918), yet right after, it is read as NUL via mmap, for no apparent reason.

**Comment by danielrainer** [comment]
`mmap`ing the history file seems quite sketchy given the unspecified behavior upon modification of the file. Are there performance-related or other issues preventing us from just reading the file into a `Vec<u8>`, e.g. via `File::read_to_end()`?

**Comment by krobelus** [comment]
On Tue, Apr 15, 2025 at 07:53:09AM -0700, danielrainer wrote:
> danielrainer left a comment (fish-shell/fish-shell#10300)
> 
> `mmap`ing the history file seems quite sketchy given the unspecified
> behavior upon modification of the file. Are there performance-related
> or other issues preventing us from just reading the file into a
> `Vec<u8>`, e.g. via `File::read_to_end()`?

the part read by fish is immutable, see https://github.com/fish-shell/fish-shell/pull/10777#issuecomment-2409263025

We should try to find a simple reproducer, I'm sure it's possible.
The problem is that this may depend on the file system (seen on at least BTRFS) and other system-specifics we don't know.

A script that runs "history append" and perhaps "history save" in a loop might be able to reproduce this.

There is a potential issue due to our usage of mtime [explained here](https://github.com/fish-shell/fish-shell/commit/5db0bd5874e76eb37f8efc22230479e6639681f7) (though that fix was broken).
So maybe inject a poor mtime granularity.

**Comment by krobelus** [comment]
This should be handled by our mtime check

**Comment by krobelus** [comment]
can_replace_file

**Comment by danielrainer** [comment]
Ah yes, then I don't see a way of having any race condition there, unless it is possible to modify the file without changing its inode, mtime, and ctime by the time they are obtained for `new_file_id`.

**Comment by radupotop** [comment]
If I may chip in, I have not noticed this issue manifesting after adding `history merge` to config.fish.  
Whether this influences the issue in any way I do not know.

**Comment by radupotop** [comment]
Apologies, ignore that, it happens regardless of `history merge`.

**Comment by danielrainer** [comment]
I just observed the issue again after about two months without it, this time on a fairly recent commit (from the past couple of days), although I'm not sure which one exactly, because I was using multiple versions concurrently. It might have been a debug build. The corruption again happened immediately after a shutdown command.

Here's the relevant xxd output from the corrupted section:
```xxd
00000000: 2d20 636d 643a 2073 6875 7464 6f77 6e20  - cmd: shutdown 
00000010: 6e6f 770a 2020 7768 656e 3a20 3137 3437  now.  when: 1747
00000020: 3039 3537 3833 0a00 0000 0000 0000 0000  095783..........
00000030: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000040: 0000 0000 0000 0000 2d20 636d 643a 2073  ........- cmd: s
00000050: 7564 6f20 7061 636d 616e 202d 5379 7975  udo pacman -Syyu
00000060: 0a20 2077 6865 6e3a 2031 3734 3731 3434  .  when: 1747144
00000070: 3438 350a 2d20 636d 643a 2063 6420 2e6c  485.- cmd: cd .l
```

**Comment by danielrainer** [comment]
A reboot triggered it again. I did not manually fix the history file. The old corruption got removed from the file but I now got

```xxd
00000000: 2d20 636d 643a 2072 6562 6f6f 740a 2020  - cmd: reboot.  
00000010: 7768 656e 3a20 3137 3437 3134 3539 3636  when: 1747145966
00000020: 0a00 0000 0000 0000 0000 0000 0000 0000  ................
00000030: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000040: 0000 2d20 636d 643a 2063 6420 2e6c 6f63  ..- cmd: cd .loc
00000050: 616c 2f73 6861 7265 2f66 6973 682f 0a20  al/share/fish/. 
```
this time definitely on `fish, version 4.0.2-913-g04fd697ac`.

**Comment by danielrainer** [comment]
After two more reboots, I got new corruption both times again, and it the penultimate sequence of NUL bytes has not been eliminated this time. Probably because only the `append` and not the `rewrite` code path were executed.

```xxd
00000000: 2d20 636d 643a 2072 6562 6f6f 740a 2020  - cmd: reboot.  
00000010: 7768 656e 3a20 3137 3437 3134 3633 3530  when: 1747146350
00000020: 0a00 0000 0000 0000 0000 0000 0000 0000  ................
00000030: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000040: 0000 2d20 636d 643a 2063 6420 2e6c 6f63  ..- cmd: cd .loc
00000050: 616c 2f73 6861 7265 2f66 6973 682f 0a20  al/share/fish/. 
00000060: 2077 6865 6e3a 2031 3734 3731 3436 3430   when: 174714640
00000070: 330a 2020 7061 7468 733a 0a20 2020 202d  3.  paths:.    -
00000080: 202e 6c6f 6361 6c2f 7368 6172 652f 6669   .local/share/fi
00000090: 7368 2f0a 2d20 636d 643a 206e 7669 6d20  sh/.- cmd: nvim 
000000a0: 6669 7368 5f68 6973 746f 7279 0a20 2077  fish_history.  w
000000b0: 6865 6e3a 2031 3734 3731 3436 3431 390a  hen: 1747146419.
000000c0: 2020 7061 7468 733a 0a20 2020 202d 2066    paths:.    - f
000000d0: 6973 685f 6869 7374 6f72 790a 2d20 636d  ish_history.- cm
000000e0: 643a 2063 6420 2e6c 6f63 616c 2f73 6861  d: cd .local/sha
000000f0: 7265 2f66 6973 682f 0a20 2077 6865 6e3a  re/fish/.  when:
00000100: 2031 3734 3731 3436 3434 390a 2020 7061   1747146449.  pa
00000110: 7468 733a 0a20 2020 202d 202e 6c6f 6361  ths:.    - .loca
00000120: 6c2f 7368 6172 652f 6669 7368 2f0a 2d20  l/share/fish/.- 
00000130: 636d 643a 2063 7020 6669 7368 5f68 6973  cmd: cp fish_his
00000140: 746f 7279 2066 6973 685f 6869 7374 6f72  tory fish_histor
00000150: 795f 636f 7272 7570 7465 645f 330a 2020  y_corrupted_3.  
00000160: 7768 656e 3a20 3137 3437 3134 3634 3531  when: 1747146451
00000170: 0a20 2070 6174 6873 3a0a 2020 2020 2d20  .  paths:.    - 
00000180: 6669 7368 5f68 6973 746f 7279 0a2d 2063  fish_history.- c
00000190: 6d64 3a20 7265 626f 6f74 0a20 2077 6865  md: reboot.  whe
000001a0: 6e3a 2031 3734 3731 3436 3439 390a 0000  n: 1747146499...
000001b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001c0: 0000 0000 0000 0000 0000 0000 0000 002d  ...............-
000001d0: 2063 6d64 3a20 6364 202e 6c6f 6361 6c2f   cmd: cd .local/
000001e0: 7368 6172 652f 6669 7368 2f0a 2020 7768  share/fish/.  wh
000001f0: 656e 3a20 3137 3437 3134 3635 3532 0a20  en: 1747146552. 
00000200: 2070 6174 6873 3a0a 2020 2020 2d20 2e6c   paths:.    - .l
00000210: 6f63 616c 2f73 6861 7265 2f66 6973 682f  ocal/share/fish/
00000220: 0a2d 2063 6d64 3a20 6e76 696d 2066 6973  .- cmd: nvim fis
00000230: 685f 6869 7374 6f72 790a 2020 7768 656e  h_history.  when
00000240: 3a20 3137 3437 3134 3635 3535 0a20 2070  : 1747146555.  p
00000250: 6174 6873 3a0a 2020 2020 2d20 6669 7368  aths:.    - fish
00000260: 5f68 6973 746f 7279 0a2d 2063 6d64 3a20  _history.- cmd: 
00000270: 6364 202e 6c6f 6361 6c2f 7368 6172 652f  cd .local/share/
00000280: 6669 7368 2f0a 2020 7768 656e 3a20 3137  fish/.  when: 17
00000290: 3437 3134 3635 3635 0a20 2070 6174 6873  47146565.  paths
000002a0: 3a0a 2020 2020 2d20 2e6c 6f63 616c 2f73  :.    - .local/s
000002b0: 6861 7265 2f66 6973 682f 0a2d 2063 6d64  hare/fish/.- cmd
000002c0: 3a20 6370 2066 6973 685f 6869 7374 6f72  : cp fish_histor
000002d0: 7920 6669 7368 5f68 6973 746f 7279 5f63  y fish_history_c
000002e0: 6f72 7275 7074 6564 5f34 0a20 2077 6865  orrupted_4.  whe
000002f0: 6e3a 2031 3734 3731 3436 3536 370a 2020  n: 1747146567.  
00000300: 7061 7468 733a 0a20 2020 202d 2066 6973  paths:.    - fis
00000310: 685f 6869 7374 6f72 790a                 h_history.
```

**Comment by jbrodriguez** [comment]
it happens to me on fish 4 too, i open a lot of terminal windows (either on ghostty or wezterm on macos), and it consistently suggests VERY old commands, that i haven't used in ages

my solution is `history merge`, after that it works normally

**Comment by danielrainer** [comment]
> my solution is history merge, after that it works normally

Yes, it's simple to fix the symptoms, but I'd like to find out what's going wrong and how to properly fix it.

**Comment by danielrainer** [comment]
I also noticed that there is a left-over temporary history file from around the time of the `shutdown` command. This file is empty.

**Comment by danielrainer** [comment]
https://github.com/fish-shell/fish-shell/pull/11492 is merged now, which should prevent all corruption of the history file on non-remote filesystems.

On remote filesystems, the types of corruption possible should be limited to losing updates. The reason for this is that we do not have a fully reliable way of ensuring exclusive access to files on remote filesystems where we cannot rely on `flock`. We address this by checking metadata of the history file before reading it, then writing the updated version to a temporary file. Once this is done, we check the metadata of the history file again, and if it is unchanged, the file is replaced by the temporary file via `rename`. The window between obtaining the metadata and `rename` completing allows for contention which would result in the version of the last writer becoming the new history file, with the updates from the other concurrent writers being lost. If we detect a metadata change, we retry by reading the updated history file and trying to update it as described previously.

Corruption other than update loss, such as the stray NUL-bytes observed by several people, should no longer occur, regardless of filesystem remoteness.

**Comment by danielrainer** [comment]
This seems to have been auto-closed by a misinterpretation of the commit message of 0f75df7c35f8b385cea3f0a9b9b3cce348d2304e, because it contains `fix #10300` in the context of `The motivation was to potentially fix #10300 which didn't succeed`.

I think closing the issue is fine, but to avoid confusion, the relevant changes are the ones in https://github.com/fish-shell/fish-shell/pull/11492, primarily f438e80f9b9dcd6a7d31e60582c35f02ef6569c5.

**Comment by danielrainer** [comment]
I just experienced history corruption following a shutdown command again, using a build of master from a few days ago. It looks the same as the previous instances; the entry for the shutdown command is intact, followed by a sequence of NUL bytes (33 this time). This happened on a local filesystem (btrfs), so the new code should ensure that there are no races between fish instances. The shutdown command was issued from a TTY with no other fish processes running. In fact, there was no other fish instance running from startup to shutdown, so I think we can safely rule out concurrent writes of multiple fish instances causing this problem.

Given that the corruption has always occurred directly after a shutdown command for me, it seems likely that there is some issue with fish's exit handling when shutting down.

**Comment by hamoid** [comment]
For me it happened the other day, unrelated to shutdown. I think I had one terminal open, opened a second one and I immediately started typing, maybe before it was ready? Using fish 4.0.2.

**Comment by danielrainer** [comment]
The prevention of race conditions described in https://github.com/fish-shell/fish-shell/issues/10300#issuecomment-2988413297 has not made it into a release yet, so in 4.0.2 these might still happen. If you experience the issue repeatedly, it might be worth checking if it disappears on a build of current master. Alternatively you can of course just wait for a release which includes the relevant changes. We still don't have a way of reliably reproducing any type of corruption mentioned in connection to this issue, which makes debugging quite challenging.

**Comment by danielrainer** [comment]
I looked at 6 corruption events for which I saved the history file. All of them are of the same style, with exactly 33 NUL bytes. Some of the commands were `shutdown now` and others `reboot`, so the entries don't have the same length. The entry lengths without NUL bytes are 39 vs. 33, respectively, or 72 vs. 64 including the NUL bytes. I checked  what the capacity of the buffer `Vec`s involved in writing to the history file would be, and AFAICT they would be 76 and 33 bytes, respectively. Given that this behavior already occurred in C++ versions, it seems unlikely that the issue is Rust-specific.

The alignment of the NUL bytes in the history file differs between corruption events and seems irrelevant.

**Comment by krobelus** [comment]
> Given that the corruption has always occurred directly after a
> shutdown command for me, it seems likely that there is some issue
> with fish's exit handling when shutting down.

maybe we're somehow not closing the file?
We should be using CLOEXEC everywhere (even though that's not relevant yet)

For extra paranoia, maybe try mounting ~/.local/share/fish on ext4 instead of btrfs.

**Comment by rpodgorny** [comment]
well, maybe it's the other way around - if this is only observed on btrfs, maybe we're up to some obscure btrfs bug...

**Comment by danielrainer** [comment]
https://github.com/fish-shell/fish-shell/issues/10300#issuecomment-2567063654 addresses issues which I think are worth exploring. We could ensure that we always `fsync` and close the file before renaming. While I don't see any indication that crashes have been involved in the corruption, it is still valuable to implement this and get rid of at least one possibility for failure.

Corruption issues have been observed on at least:

- apfs
- btrfs
- ext4
- zfs

Some of these might have been due to fish instances racing on the history file, but it is unclear if that is the case. If we get out a release with these changes, in addition to the previous ones from https://github.com/fish-shell/fish-shell/pull/11492, we will get more information about whether corruption still occurs then. It would probably help to reopen this issue to make it easier to find.

**Comment by pvonmoradi** [comment]
Thanks @krobelus for reoppening, for the record here is a screencast of me experiencing this issue in Debian 13, fish `v4.0.2`: https://i.imgur.com/LeCAjVS.mp4

This happens after booting from a `sudo reboot`

**Comment by mlncn** [comment]
As far as workarounds, is anyone automatically deleting null values, and/or is there a recommended `history merge` automation that fixes this?  The one big fish frustration.  NixOs, ext4 on an SSD, fish 4.0.2 here (but have also had the problem for a while).  Use Konsole and i do indeed have a ridiculous number (like often 50 or more) windows/tabs open at once with lots of switching, creating and closing, etc.

**Comment by danielrainer** [comment]
I don't have any automation to work around the issue, but since I've been using the latest changes relating to this issue (early August), I have not had any corruption, which makes me cautiously optimistic that the issue is fixed (at least in some cases). The relevant changes have not made it into a release yet, but if you are comfortable with using a development build, you could check if it fixes the corruption for you. Having more testing would also be helpful to increase confidence that the issue is really fixed, since we haven't found a way to reliably reproduce these corruption events.

**Comment by mithereal** [comment]
this issue has gotten so bad i went back to bash after a decade. you had a great working product but all yall do is tweak everything, i hate computers now, its like trying to be successful while a methhead keeps messing everyting up.

**Comment by mithereal** [comment]
im having this issue, however its a lil different i only have one console open this was never an issue for me until about 3 month ago., it just gets worse and worse as time goes on ive used fish prolly for around 10 yrs, but now its an unusable product, i went back to 3.0.1 and the issue doesnt exist.

**Comment by awsms** [comment]
I'm definitely still having this issue on 4.7.1, as I've been having it for years. Why was this issue closed?

**Comment by danielrainer** [comment]
Several problems have been identified and fixed. If you still have issues with history corruption on recent builds of fish it would be helpful to know what exact symptoms you observe, which operating system you use, and which file system the history file is stored on.

**Comment by magnetophon** [comment]
I also occasionally still run into this, using 4.7.1.
The symptoms are the same as in the initial bug report except it's on different terminals instead of different tmux windows.
I use NixOS unstable and the history file is stored on ZFS.

``history merge`` fixes it.

**Comment by theodore-s-beers** [comment]
I experience this issue multiple times a week, most recently today. fish 4.7.1 in Ghostty on macOS.

edit:

I think in my case, this may be an unfortunate interaction of two deliberate fish behaviors – each session filters file-backed history through its fixed `boundary_timestamp`, while periodic history rewriting/vacuuming deduplicates repeated commands and can leave only a newer record for a frequently run command.

A pre-existing session that reloads the rewritten history still applies its original boundary, so it can hide the newer "good" command and fall back to an older typo. `history merge` fixes this because it calls `incorporate_external_changes()`, which advances the boundary timestamp and clears the cached history file state.

Disclaimer – I am new to the fish codebase and could be wrong about any/every part of this.

**Comment by theodore-s-beers** [comment]
Sorry to leave a second comment in a row, but I wanted to demonstrate how I can get fish to give me an autocomplete suggestion of an older typo, due to the way the timestamp of a frequently run command gets updated on deduplication. It then falls after the boundary of a pre-existing session and is thus invisible to it. An older typo surfaces instead. All of this behavior may be intentional, but understanding it now is helpful to me.

```
--- SESSION I WILL INTENTIONALLY BREAK ---

Last login: Wed Jun 10 13:07:59 on ttys023

fish-shell
❯ set good 'jj git fetch'

fish-shell
❯ set bad 'jj git fexit'

fish-shell
❯ history search --exact --case-sensitive $good --show-time='%F %T'
2026-06-10 13:06:18jj git fetch

fish-shell
❯ history search --exact --case-sensitive $bad --show-time='%F %T'
2026-06-09 11:16:42jj git fexit


--- SEPARATE SESSION I USE TO BREAK THE FIRST ONE ---

Last login: Wed Jun 10 13:05:13 on ttys022

fish-shell
❯ history delete --exact --case-sensitive 'jj git fetch'

fish-shell
❯ history append 'jj git fetch'

fish-shell
❯ history save


--- BACK IN THE NOW-BROKEN SESSION ---

fish-shell
❯ true

fish-shell
❯ history search --exact --case-sensitive $good --show-time='%F %T'

fish-shell
❯ history search --exact --case-sensitive $bad --show-time='%F %T'
2026-06-09 11:16:42jj git fexit

fish-shell
❯ history merge

fish-shell
❯ history search --exact --case-sensitive $good --show-time='%F %T'
2026-06-10 13:09:31jj git fetch
```

**Comment by radupotop** [comment]
Does it help if you add `history merge` at the bottom of your `config.fish`?

**Comment by magnetophon** [comment]
> Does it help if you add `history merge` at the bottom of your `config.fish`?

That would only run when starting the terminal, right?
So then it would not help; the problem occurs after I've been running the terminal for a while.

**Comment by magnetophon** [comment]
@krobelus IMHO, this issue should be re-opened.

**Comment by theodore-s-beers** [comment]
> That would only run when starting the terminal, right?

Correct. And it's not even needed at session start. Let me explain what I'm doing instead. I forked fish to add an `EPOCHSECONDS` variable, allowing for very cheap lookup of "What time is it now?" (I'm seeing ~10 μs on that.)

Then I added this function to my fish config, which runs on every prompt and does `history merge` at most every 60 seconds. I may adjust that threshold in the future. But what was important to me was a) that `history merge` not run on every prompt; and b) that the mechanism of throttling how often it runs have nearly zero cost.

```fish
if status is-interactive
    function __maybe_history_merge --on-event fish_prompt
        set -l now $EPOCHSECONDS

        if not set -q __next_history_merge
            set -g __next_history_merge (math $now + 60)
        end

        if test $now -ge $__next_history_merge
            if history merge
                set -g __next_history_merge (math $now + 60)
            end
        end
    end
end
```

**Comment by danielrainer** [comment]
I can confirm that [this](https://github.com/fish-shell/fish-shell/issues/10300#issuecomment-4672643095) reproduces, assuming you start by putting both the "good" and "bad" command into the history first. This is a valuable observation, because so far we've not had any way of reliably reproducing issues with the history.

I looked into the code and there is indeed an issue with timestamps not being refreshed on reloading the history file. In particular `src/history/history.rs:load_old_if_needed` loads the current version of the history file and then calls `history_file.decode(Some(self.boundary_timestamp))` without updating the timestamp. Some levels further down in the call stack, this reaches `src/history/yaml_backend.rs:offset_of_next_item_fish_2_0`, where history items with a timestamp that's more recent than the timestamp that was passed are ignored. The relevant code contains this comment:
```text
Hackish fast way to skip items created after our timestamp. This is the mechanism by
which we avoid "seeing" commands from other sessions that started after we started.
```
This explains the behavior triggered by your repro. Deleting the "good" entry adds it to the `deleted_items` map at first, without modifying the history file. Then, when the command is added to the history again, either via `history append`, or by simply entering the command, the history file is written. Usually, in cases where only new entries are added, we would append to the file, but because we have a deleted item, `self.save_internal_via_rewrite` is triggered in `src/history/history.rs:HistoryImpl::save`. This rewrites the entire file, removing the old entry of the "good" command. Triggering completions in a different fish then calls `src/history/history.rs:load_old_if_needed`, but there the timestamp has not been updated, so when reloading it ignores the new "good" entry, so it does not appear in the history.

A simplified repro:
1. Have some command in the history, e.g. `echo asdf`.
2. Open 2 fish.
3. Run `history delete --exact --case-sensitive 'echo asdf'` and `echo asdf` in one fish.
4. Try getting completions/history for `echo asdf` in the other fish. It won't show up.

Note that this can happen without deleting the command. It is also possible to enter that command, causing it to be added to the `new_items` Vec of `HistoryImpl` with a fresh timestamp. Then, the history file needs to be rewritten by the same fish where the command was entered, which can happen in one of several ways. Options include deleting some other history item and entering another command, or entering enough commands that the randomly initialized `VACUUM_FREQUENCY` is exceeded. At most 25 distinct commands are needed for the latter.

I'm not yet sure how to best address this, but I agree that it is an issue that should be fixed.

**Comment by theodore-s-beers** [comment]
> Note that this can happen without deleting the command. It is also possible to enter that command, causing it to be added to the `new_items` Vec of `HistoryImpl` with a fresh timestamp. Then, the history file needs to be rewritten by the same fish where the command was entered, which can happen in one of several ways. Options include deleting some other history item and entering another command, or entering enough commands that the randomly initialized `VACUUM_FREQUENCY` is exceeded. At most 25 distinct commands are needed for the latter.

Yes, I was hoping that someone familiar with the fish codebase would recognize that there are ways for this to happen organically.

**Comment by jbrodriguez** [comment]
> I'm not yet sure how to best address this

a global lock, the python gil 🤣
